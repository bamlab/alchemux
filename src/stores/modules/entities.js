// @flow
import { mergeDeepRight, path } from 'ramda';

type ID = string | number;
type Entities = {
  [ID]: Object,
};

type State = {
  [string]: Entities,
};

type StoreKeyType = 'entities';
export const STORE_KEY: StoreKeyType = 'entities';
export type GlobalState = { [StoreKeyType]: State };

type Action = {
  type: string,
  meta?: {
    entities?: {
      [string]: Entities,
    },
  },
};

export const addEntitiesToAction = <A>(action: A, entities: { [string]: Entities }): A & Action => {
  return mergeDeepRight(action, {
    meta: {
      entities,
    },
  });
};

const initialState = {};

export default function reducer(state: State = initialState, action: Action) {
  const entities = path(['meta', 'entities'], action);
  if (!entities) {
    return state;
  }
  return mergeDeepRight(state, entities);
}

export const entitiesSelector = (state: GlobalState, entityName: string): ?Entities =>
  state[STORE_KEY][entityName];
