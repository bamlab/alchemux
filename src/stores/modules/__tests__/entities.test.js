import { default as reducer, entitiesSelector, STORE_KEY } from '../entities';

describe('entities', () => {
  describe('reducer', () => {
    it('should have an initial state', () => {
      expect(reducer(undefined, { type: '@INIT' })).toMatchSnapshot();
    });
    it('should store given entities', () => {
      const action = {
        type: 'Anything',
        meta: {
          entities: {
            user: {
              3: { id: 3, name: 'foo' },
              5: { id: 5, name: 'bar' },
            },
            comment: {
              uid1: { id: 'uid1', text: 'barfoo' },
              uid2: { id: 'uid2', text: 'foorbar' },
            },
          },
        },
      };
      expect(reducer(undefined, action)).toMatchSnapshot();
    });
    it('should update the new entities', () => {
      const action = {
        type: 'Anything',
        meta: {
          entities: {
            user: {
              3: { id: 3, name: 'foo' },
              5: { id: 5, name: 'bar' },
            },
            comment: {
              uid1: { id: 'uid1', text: 'barfoo' },
            },
          },
        },
      };
      const updateAction = {
        type: 'Anything',
        meta: {
          entities: {
            user: {
              3: { id: 3, name: 'foobar_barfoo' },
            },
          },
        },
      };
      const state = reducer(undefined, action);
      expect(reducer(state, updateAction)).toMatchSnapshot();
    });
  });
});
