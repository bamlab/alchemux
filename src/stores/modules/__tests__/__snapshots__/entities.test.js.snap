// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entities reducer should have an initial state 1`] = `Object {}`;

exports[`entities reducer should store given entities 1`] = `
Object {
  "comment": Object {
    "uid1": Object {
      "id": "uid1",
      "text": "barfoo",
    },
    "uid2": Object {
      "id": "uid2",
      "text": "foorbar",
    },
  },
  "user": Object {
    "3": Object {
      "id": 3,
      "name": "foo",
    },
    "5": Object {
      "id": 5,
      "name": "bar",
    },
  },
}
`;

exports[`entities reducer should update the new entities 1`] = `
Object {
  "comment": Object {
    "uid1": Object {
      "id": "uid1",
      "text": "barfoo",
    },
  },
  "user": Object {
    "3": Object {
      "id": 3,
      "name": "foobar_barfoo",
    },
    "5": Object {
      "id": 5,
      "name": "bar",
    },
  },
}
`;
